// classical-cipher/dist/Keys/FixedArrangement.js
"use strict";
var _ = require("lodash");
var keysUtils_1 = require("./keysUtils");
var util_1 = require("../util");
var InvalidKeyException_1 = require("./InvalidKeyException");
var FixedArrangement = (function () {
    function FixedArrangement(alphabet) {
        this.alphabet = alphabet;
    }
    FixedArrangement.prototype.startGeneration = function () {
        return this.alphabet;
    };
    FixedArrangement.prototype.switchLetters = function (arr) {
        var letter1Loc = _.random(0, arr.length - 1);
        var letter2Loc = _.random(0, arr.length - 1);
        var tmp = arr[letter1Loc];
        arr[letter1Loc] = arr[letter2Loc];
        arr[letter2Loc] = tmp;
        return arr;
    };
    FixedArrangement.prototype.smallAlteration = function (arr) {
        return this.switchLetters(arr);
    };
    Object.defineProperty(FixedArrangement.prototype, "range", {
        get: function () {
            return util_1.factorial(this.alphabet.length);
        },
        enumerable: true,
        configurable: true
    });
    FixedArrangement.prototype.getPrimitiveKey = function (key) {
        if (_.uniq(key).length == key.length && typeof key === "object") {
            return key;
        }
        else {
            throw new InvalidKeyException_1.default(key);
        }
    };
    FixedArrangement.prototype.generateRandom = function () {
        return _.shuffle(this.alphabet);
    };
    FixedArrangement.prototype.generateOrdered = function (previous) {
        return keysUtils_1.nextPermutation(previous);
    };
    return FixedArrangement;
}());
module.exports = FixedArrangement;
