// classical-cipher/dist/Ciphers/Cipher.js
"use strict";
var _ = require("lodash");
var Solver_1 = require("../Solvers/Solver");
var util_1 = require("../util");
var EncyptionOptionsDefault = {
    formatResult: true
};
var Cipher = (function () {
    function Cipher() {
        Cipher.allCiphers.push(this);
    }
    Cipher.prototype.formattedCipherOperation = function (input, key, options, operation) {
        _.defaults(options, EncyptionOptionsDefault);
        var decryptedText = operation(util_1.unformat(input), this.keyInfo.getPrimitiveKey(key));
        return options.formatResult ? util_1.format(decryptedText, input) : decryptedText;
    };
    Cipher.prototype.encrypt = function (input, key, options) {
        if (options === void 0) { options = {}; }
        return this.formattedCipherOperation(input, key, options, this.rawEncrypt.bind(this));
    };
    Cipher.prototype.decrypt = function (input, key, options) {
        if (options === void 0) { options = {}; }
        return this.formattedCipherOperation(input, key, options, this.rawDecrypt.bind(this));
    };
    Cipher.prototype.solve = function (args) {
        args.cipher = this;
        return Solver_1.solve(args);
    };
    ;
    Cipher.allCiphers = [];
    return Cipher;
}());
exports.Cipher = Cipher;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Cipher;
