// classical-cipher/dist/Reporters/SilentReporter.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Reporter = require("./Reporter");
var SilentReporter = (function (_super) {
    __extends(SilentReporter, _super);
    function SilentReporter() {
        _super.apply(this, arguments);
    }
    SilentReporter.prototype.setOriginalText = function (originalText) {
    };
    SilentReporter.prototype.start = function (numIncrement) {
    };
    SilentReporter.prototype.update = function (toWhat) {
        if (toWhat === void 0) { toWhat = 7; }
    };
    SilentReporter.prototype.finish = function (plainText, key, strength) {
    };
    SilentReporter.prototype.log = function (text) {
    };
    SilentReporter.prototype.error = function (text) {
    };
    return SilentReporter;
}(Reporter));
module.exports = SilentReporter;
