// classical-cipher/dist/Reporters/Stdout.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Reporter = require("./Reporter");
var Stdout = (function (_super) {
    __extends(Stdout, _super);
    function Stdout() {
        _super.apply(this, arguments);
        this.incrementOn = 0;
    }
    Stdout.prototype.setOriginalText = function (originalText) {
        this.originalText = originalText;
    };
    Stdout.prototype.start = function (numIncrement) {
        this.numIncrement = numIncrement;
    };
    Stdout.prototype.update = function (toWhat) {
        if (toWhat === void 0) { toWhat = ++this.incrementOn; }
        console.log("Completed " + Math.floor((toWhat / this.numIncrement) * 100) + "%");
    };
    Stdout.prototype.finish = function (plainText, key, strength) {
        console.log("Task Finished");
        console.log(plainText);
        console.log("Strength: " + strength);
        console.log("Key: " + key);
    };
    Stdout.prototype.log = function (text) {
        console.log(text);
    };
    Stdout.prototype.error = function (text) {
        console.error(text);
    };
    return Stdout;
}(Reporter));
module.exports = Stdout;
